### ISSUE TITLE
feat(training): Add WorkerConsistencyValidator to fix Issue #562 pattern (Issue #274 enhancement)

### ISSUE THREAD

**Issue by akaszubski** [akaszubski/autonomous-dev#281]
## Summary

Add WorkerConsistencyValidator library to ensure all distributed workers use identical prompts and code, preventing inconsistent scoring. Updates distributed-training-coordinator agent with pre-training validation.

## What Does NOT Work

**Pattern: Workers use different prompts causing inconsistent results (Issue #562)**
- ReAlign bug: Local used full Tulu3 prompt (1500 chars), remote used compact prompt (400 chars)
- Same example got different quality scores depending on which worker processed it
- Result: Training data quality inconsistent, model learns from biased data
- Current distributed-training-coordinator has no worker validation

**Pattern: No verification that workers loaded same code**
- Workers may have different agent versions if codebase not synced
- Supporting modules (scoring prompts, quality thresholds) may differ
- No check that model checkpoint is identical across workers
- Gradients synced from inconsistent starting states

### Fresh Install
**What happens**: User installs Issue #274 training components
- distributed-training-coordinator starts workers without validation
- No check that workers use identical code/prompts

**Expected behavior after this issue**:
- WorkerConsistencyValidator.validate_workers() runs before training
- Detects prompt/code mismatches
- Fails fast with clear error message

### Update Existing Issue #274 Installation
**What happens**: User has distributed-training-coordinator already
- Upgrade adds worker validation step

**Expected behavior**:
- Agent enhanced with Phase 2.5: Worker Consistency Validation
- Existing workflows preserved (validation optional for simple cases)
- Validation results logged to audit trail

### Distributed Training with Worker Mismatch (Caught Early)
**What happens**: Local has updated Tulu3 prompt, remote has old version
- Old behavior: Training proceeds, produces inconsistent scores
- New behavior: WorkerConsistencyValidator detects mismatch, fails before training

**Expected behavior**:
- Validation detects prompt mismatch (hash comparison)
- Clear error: "Worker 'remote' has different tulu3_scoring_prompt.py (hash: abc123 vs expected: def456)"
- Provides fix command: "Run ~/Dev/sync-dev.sh to sync codebase"

### 1. Create WorkerConsistencyValidator Library

**File**: `plugins/autonomous-dev/lib/worker_consistency_validator.py` (NEW)

```python
class WorkerConsistencyValidator:
    """Validate all distributed workers use identical code and prompts.
    
    Issue: Enhancement to Issue #274 distributed-training-coordinator
    Source: ReAlign Issue #562 (worker prompt consistency bug)
    Pattern: Tulu3 uses sum loss instead of mean loss for cross-device averaging
    """
    
    def validate_workers(
        self,
        workers: List[str],
        critical_files: List[str]
    ) -> ValidationResult:
        """Validate workers have identical code files.
        
        Args:
            workers: List of worker IDs
            critical_files: Files that must match (e.g., scoring_prompt.py)
        
        Returns:
            ValidationResult with pass/fail and mismatch details
        """
        file_hashes = {}
        for worker in workers:
            worker_hashes = compute_file_hashes(worker, critical_files)
            file_hashes[worker] = worker_hashes
        
        # Check all workers match
        reference_hashes = file_hashes[workers[0]]
        mismatches = []
        for worker in workers[1:]:
            for file, hash_val in file_hashes[worker].items():
                if hash_val != reference_hashes[file]:
                    mismatches.append({
                        "worker": worker,
                        "file": file,
                        "expected": reference_hashes[file],
                        "actual": hash_val
                    })
        
        return ValidationResult(
            passed=len(mismatches) == 0,
            mismatches=mismatches
        )
    
    def validate_model_checkpoint(
        self,
        workers: List[str],
        checkpoint_path: str
    ) -> bool:
        """Verify all workers loaded same model checkpoint."""
        checksums = {}
        for worker in workers:
            checksums[worker] = compute_checkpoint_hash(worker, checkpoint_path)
        
        return len(set(checksums.values())) == 1
    
    def validate_random_seed(
        self,
        workers: List[str]
    ) -> bool:
        """Verify all workers set identical random seeds."""
        seeds = {}
        for worker in workers:
            seeds[worker] = get_worker_random_state(worker)
        
        return len(set(seeds.values())) == 1
```

### 2. Update distributed-training-coordinator Agent

**File**: `plugins/autonomous-dev/agents/distributed-training-coordinator.md` (UPDATE)

Add Phase 2.5: Worker Consistency Validation (after Phase 2: Work Distribution, before Phase 3: Gradient Sync):

```markdown
## Phase 2.5: Worker Consistency Validation (CRITICAL)

**Purpose**: Prevent Issue #562 pattern (inconsistent worker code/prompts)

**Steps**:
1. Create WorkerConsistencyValidator instance
2. Validate critical files match across all workers:
   - Scoring prompts (e.g., tulu3_scoring_prompt.py)
   - Quality thresholds (content_type_quality_thresholds.py)
   - Agent code (distributed-training-coordinator itself)
3. Validate model checkpoint loaded identically on all workers
4. Validate random seeds set identically (numpy, mlx)
5. If validation fails:
   - Log mismatch details to audit log
   - Display fix command: "Run ~/Dev/sync-dev.sh --no-confirm"
   - Abort training (do not proceed to Phase 3)

**Tulu3 Pattern**: Use sum loss instead of mean loss for cross-device averaging
- Ensures consistent gradient aggregation despite worker differences
- mlx.nn.average_gradients() then divide by world_size
```

### 3. Update mlx-distributed.md Documentation

**File**: `plugins/autonomous-dev/skills/mlx-performance/docs/mlx-distributed.md` (UPDATE)

Add section:

```markdown
## Worker Consistency (Issue #562 Pattern)

Before training, validate all workers use identical code:

\`\`\`python
from worker_consistency_validator import WorkerConsistencyValidator

validator = WorkerConsistencyValidator()
result = validator.validate_workers(
    workers=["rank-0", "rank-1"],
    critical_files=[
        "src/realign/data/quality/tulu3_scoring_prompt.py",
        "src/realign/data/quality/content_type_thresholds.py"
    ]
)

if not result.passed:
    print(f"Worker mismatch detected:")
    for mismatch in result.mismatches:
        print(f"  {mismatch['worker']}: {mismatch['file']}")
        print(f"    Expected: {mismatch['expected']}")
        print(f"    Actual: {mismatch['actual']}")
    
    print("Fix: Run ~/Dev/sync-dev.sh --no-confirm")
    sys.exit(1)
\`\`\`

**Tulu3 Cross-Device Averaging**:
- Use sum loss (not mean loss) for consistent aggregation
- Pattern: loss_sum = dist.all_reduce(loss, "sum"); loss_mean = loss_sum / world_size
```

### Detect Worker Prompt Mismatch
- [ ] Create worker A with prompt version 1 (1500 chars)
- [ ] Create worker B with prompt version 2 (400 chars)
- [ ] Run validate_workers() with tulu3_scoring_prompt.py
- [ ] Verify validation fails with clear mismatch details
- [ ] Verify fix command provided

### Validate Identical Workers (Pass)
- [ ] Create 4 workers with identical code
- [ ] Run validate_workers() on critical files
- [ ] Verify validation passes
- [ ] Verify audit log shows "Worker consistency: PASS"

### Validate Model Checkpoint Match
- [ ] Load checkpoint on rank 0
- [ ] Load same checkpoint on rank 1, 2, 3
- [ ] Run validate_model_checkpoint()
- [ ] Verify validation passes

### Validate Random Seed Match
- [ ] Set numpy seed=42 on all workers
- [ ] Set mlx seed=42 on all workers
- [ ] Run validate_random_seed()
- [ ] Verify validation passes

### Integration with distributed-training-coordinator
- [ ] Agent runs WorkerConsistencyValidator in Phase 2.5
- [ ] Mismatch detected, training aborted
- [ ] Clear error message with fix command
- [ ] User runs sync-dev.sh, re-runs, validation passes

### New Library
- [ ] WorkerConsistencyValidator class created
- [ ] validate_workers() checks file hashes across workers
- [ ] validate_model_checkpoint() verifies checkpoint match
- [ ] validate_random_seed() verifies seed consistency
- [ ] ValidationResult includes mismatch details and fix commands

### Updates to Issue #274 Components
- [ ] distributed-training-coordinator agent enhanced with Phase 2.5
- [ ] Phase 2.5 runs after work distribution, before gradient sync
- [ ] Agent aborts training if validation fails (fail-fast)
- [ ] mlx-distributed.md documents worker consistency pattern

### Tulu3 Pattern Integration
- [ ] Documentation mentions sum loss (not mean loss) pattern
- [ ] Example shows dist.all_reduce(loss, "sum") / world_size
- [ ] Explains why: Consistent aggregation despite worker differences

### Security
- [ ] File path validation (CWE-22) for critical_files parameter
- [ ] Hash computation uses secure algorithm (SHA256)
- [ ] Audit logging for all validation results
- [ ] No secrets exposed in mismatch error messages

## Environment Requirements

- **Python**: 3.11+ (existing requirement)
- **MLX**: 0.30.4+ (for distributed support)
- **Optional**: ~/Dev/sync-dev.sh for codebase synchronization

## Source of Truth

**Verified**: Real-world ReAlign distributed MLX work
- Issue #562: Worker prompt consistency bug (fixed with this pattern)
- Local: 1500-char prompt, Remote: 400-char prompt → inconsistent scores
- Tulu3 paper: Sum loss for cross-device averaging (https://arxiv.org/abs/2411.15124)

**Date Verified**: 2026-01-28

## Related Issues

- #274 (CLOSED): Training agents baseline - enhances distributed-training-coordinator
- #562: Worker prompt consistency bug (ReAlign-specific, pattern captured here)
- #279, #280: Foundation issues in this series
- Part of 6-issue series for ReAlign distributed MLX patterns

**Estimated Implementation Time**: 2-3 hours
